From 2675f60864a4ddabb33bff282f223242ca99b792 Mon Sep 17 00:00:00 2001
From: Kleber Sacilotto de Souza <kleber.souza@canonical.com>
Date: Wed, 9 Oct 2019 11:20:58 +0200
Subject: [PATCH] UBUNTU: SAUCE: skip fanotify09 test #2 for older kernels

BugLink: https://bugs.launchpad.net/bugs/1804594

Test case #2 was added to fanotify09 as a regression test for the
following upstream commit:

b469e7e47c8a fanotify: fix handling of events on child sub-directory

which has been applied to v4.20 and backported to linux-4.19.y upstream
stable.

We have backported this change to Bionic and it was included on
4.15.0-59. The backport for 4.4.0 and older kernels are not
straightforward, so skip testcase #2 for kernels older than 4.15.0.

Signed-off-by: Kleber Sacilotto de Souza <kleber.souza@canonical.com>
---
 testcases/kernel/syscalls/fanotify/fanotify09.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/testcases/kernel/syscalls/fanotify/fanotify09.c b/testcases/kernel/syscalls/fanotify/fanotify09.c
index 83210bc1c..e9f34cb84 100644
--- a/testcases/kernel/syscalls/fanotify/fanotify09.c
+++ b/testcases/kernel/syscalls/fanotify/fanotify09.c
@@ -192,6 +192,15 @@ static void test_fanotify(unsigned int n)
 
 	tst_res(TINFO, "Test #%d: %s", n, tc->tname);
 
+	/*
+	 * The fix tested by test case #2 is not present on Ubuntu kernels
+	 * older than 4.15 (Bionic).
+	 */
+	if ((n == 1) && (tst_kvercmp(4, 15, 0) < 0)) {
+		tst_brk(TCONF, "handling of events on child sub-directory "
+			"is broken with this kernel");
+	}
+
 	create_fanotify_groups(tc);
 
 	/*
-- 
2.17.1

